code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm

```diff
@@ -141,94 +143,82 @@
+
+	if (main_port.total_moles() < 0.01)
+		skipping_work = TRUE
+		return
+
+	if(cooling)
+		var/datum/gas_mixture/exchange_target
+		// Exchange target is the thing we are paired with, be it enviroment or the red port. 
+		if(use_enviroment_heat)
+			exchange_target = local_turf.return_air()
+		else
+			exchange_target = airs[2]
+			
+		if (exchange_target.total_moles() < 0.01)
+			skipping_work = TRUE
 			return
-		if(enviroment.temperature > THERMOMACHINE_SAFE_TEMPERATURE && safeties)
+
+		// The hotter the heat reservoir is, the larger the malus.
+		var/temperature_exchange_delta = exchange_target.temperature - main_port.temperature
+		// Log 1 is already 0, going any lower will result in a negative number.
+		efficiency = clamp(1 - log(10, max(1, temperature_exchange_delta*0.08)), 0.65, 1)
+		// We take an extra efficiency malus for enviroments where the mol is too low.
+		if (use_enviroment_heat)
+			efficiency *= clamp(log(1.55, exchange_target.total_moles()) * 0.15, 0.65, 1)
+
+		if (exchange_target.temperature > THERMOMACHINE_SAFE_TEMPERATURE && safeties)
 			on = FALSE
-			visible_message("<span class='warning'>The enviroment's temperature has reached critical levels, shutting down...</span>")
+			visible_message("<span class='warning'>The heat reservoir has reached critical levels, shutting down...</span>")
 			update_appearance()
 			return
-		else if(enviroment.temperature > THERMOMACHINE_SAFE_TEMPERATURE && !safeties)
+
+		else if(exchange_target.temperature > THERMOMACHINE_SAFE_TEMPERATURE && !safeties)
 			if((REALTIMEOFDAY - lastwarning) / 5 >= WARNING_DELAY)
 				lastwarning = REALTIMEOFDAY
-				visible_message("<span class='warning'>The enviroment's temperature has reached critical levels!</span>")
-				if(check_explosion(enviroment.temperature))
+				visible_message("<span class='warning'>The heat reservoir has reached critical levels!</span>")
+				if(check_explosion(exchange_target.temperature))
 					explode()
-					return PROCESS_KILL //we dying anyway, so let's stop processing
-		temperature_difference = enviroment.temperature - main_port.temperature
-		temperature_difference = cooling ? temperature_difference : 0
-		if(temperature_difference > 0)
-			efficiency = max(1 - log(10, temperature_difference) * 0.08, 0.65)
-		main_port.temperature = max(main_port.temperature - (heat_amount * efficiency * enviroment_efficiency) / main_heat_capacity + motor_heat / main_heat_capacity, TCMB)
-		skip_tick = FALSE
+					return PROCESS_KILL //We're dying anyway, so let's stop processing
+		
+		exchange_target.temperature = max((THERMAL_ENERGY(exchange_target) - (heat_amount * efficiency) + motor_heat) / exchange_target.heat_capacity(), TCMB)
 
-	skipping_work = skip_tick
+	main_port.temperature = max((THERMAL_ENERGY(main_port) + (heat_amount * efficiency) + motor_heat) / main_port.heat_capacity(), TCMB)
 
 	heat_amount = abs(heat_amount)
 	var/power_usage = 0
-	if(temperature_delta  > 1)
+	if(abs(temperature_target_delta  > 1))
```

Comment:
i think you want to put the `temperature_target_delta` between the () of abs
`if(abs(temperature_target_delta) > 1)`

Reply:
good catch, thanks